contracts/limitOrder/LimitOrderBook.sol

```diff
@@ -0,0 +1,170 @@
+import { EIP712Upgradeable } from "@openzeppelin/contracts-upgradeable/drafts/EIP712Upgradeable.sol";
+import { SignedSafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SignedSafeMathUpgradeable.sol";
+import { PerpMath } from "@perp/curie-contract/contracts/lib/PerpMath.sol";
+import { ILimitOrderBook } from "../interface/ILimitOrderBook.sol";
+import { ILimitOrderFeeVault } from "../interface/ILimitOrderFeeVault.sol";
+import { LimitOrderBookStorageV1 } from "../storage/LimitOrderBookStorage.sol";
+import { OwnerPausable } from "../base/OwnerPausable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import { IClearingHouse } from "@perp/curie-contract/contracts/interface/IClearingHouse.sol";
+import { IAccountBalance } from "@perp/curie-contract/contracts/interface/IAccountBalance.sol";
+
+contract LimitOrderBook is
+    ILimitOrderBook,
+    BlockContext,
+    ReentrancyGuardUpgradeable,
+    OwnerPausable,
+    EIP712Upgradeable,
+    LimitOrderBookStorageV1
+{
+    using AddressUpgradeable for address;
+    using PerpMath for int256;
+    using PerpMath for uint256;
+    using SignedSafeMathUpgradeable for int256;
+
+    // NOTE: cannot use `OrderType orderType` here, use `uint256 orderType` instead
+    // solhint-disable-next-line func-name-mixedcase
+    bytes32 public constant LIMIT_ORDER_TYPEHASH =
+        keccak256(
+            // solhint-disable-next-line max-line-length
+            "LimitOrder(uint256 orderType,uint256 salt,address trader,address baseToken,bool isBaseToQuote,bool isExactInput,uint256 amount,uint256 oppositeAmountBound,uint256 deadline,bool reduceOnly,uint80 roundIdWhenCreated,uint256 triggerPrice)"
+        );
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    function initialize(
+        string memory name,
+        string memory version,
+        address clearingHouseArg,
+        address limitOrderFeeVaultArg
+    ) external initializer {
+        __ReentrancyGuard_init();
+        __OwnerPausable_init();
+        __EIP712_init(name, version); // ex: "PerpCurieLimitOrder" and "1"
+
+        // LOB_CHINC : ClearingHouse Is Not Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
+
+        // LOB_LOFVINC : LimitOrderFeeVault Is Not Contract
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+    }
+
+    function setClearingHouse(address clearingHouseArg) external onlyOwner {
+        // LOB_CHINC: ClearingHouse Is Not a Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
```

Comment:
better to check `accountBalance` is not null. It might happen when `clearingHouse` is a contract but not our CH

Reply:
@vinta the checking should be included in `setClearingHouse` as well, not only `initialize`